packages/cbioportal-utils/src/ptm/PtmUtils.ts

```diff
@@ -0,0 +1,377 @@
+    'Pyruvic acid (Cys)': PtmType.Pyruvate,
+    'Pyruvic acid (Ser)': PtmType.Pyruvate,
+    'Pyruvic acid (Tyr)': PtmType.Pyruvate,
+    // S-nitrosylation
+    'S-nitrosocysteine': PtmType.SNitrosylation,
+    // Sulfation
+    Sulfoserine: PtmType.Sulfation,
+    Sulfothreonine: PtmType.Sulfation,
+    Sulfotyrosine: PtmType.Sulfation,
+};
+
+export function groupPtmDataByPosition(
+    ptmData: PostTranslationalModification[]
+) {
+    return _.groupBy(ptmData, p => p.residue.start);
+}
+
+export function groupPtmDataByTypeAndPosition(
+    ptmData: PostTranslationalModification[]
+) {
+    const groupedByType = _.groupBy(ptmData, p => p.type);
+
+    const groupedByTypeAndPosition: {
+        [type: string]: { [position: number]: PostTranslationalModification[] };
+    } = {};
+
+    _.keys(groupedByType).forEach(type => {
+        groupedByTypeAndPosition[type] = _.groupBy(
+            groupedByType[type],
+            p => p.residue.start
+        );
+    });
+
+    return groupedByTypeAndPosition;
+}
+
+export function ptmColor(ptms: PostTranslationalModification[]) {
+    let color = PTM_COLORS.default;
+    const uniqueTypes = _.uniq((ptms || []).map(ptm => ptm.type));
+
+    if (uniqueTypes.length === 1) {
+        color = PTM_COLORS[uniqueTypes[0]] || PTM_COLORS.default;
+    } else if (uniqueTypes.length > 1) {
+        color = PTM_COLORS.multiType;
+    }
+
+    return color;
+}
+
+export function compareByPtmTypePriority(a: string, b: string) {
+    const aValue = PTM_PRIORITY[a] || PTM_PRIORITY.default;
+    const bValue = PTM_PRIORITY[b] || PTM_PRIORITY.default;
+
+    if (aValue > bValue) {
+        return 1;
+    } else if (bValue > aValue) {
+        return -1;
+    } else if (a > b) {
+        return 1;
+    } else if (b > a) {
+        return -1;
+    }
+
+    return 0;
+}
+
+function getResiduesFromUniportFeature(ptm: UniprotFeature) {
+    return {
+        start: Number(ptm.begin),
+        end: Number(ptm.end),
+    };
+}
+
+export function getPubmedIdsFromUniportFeature(ptm: UniprotFeature) {
+    return ptm.evidences
+        .filter(
+            e =>
+                e.source &&
+                e.source.name &&
+                e.source.name.toLowerCase().includes('pubmed')
+        )
+        .map(e => e.source.id);
+}
+
+export function getPtmTypeFromUniportFeature(ptm: UniprotFeature) {
+    // actual `type` field provided by the feature itself (MOD_RES, CROSS_LINK, etc.) is not what we want to display
+    // we need specific PTM types (Ubiquitination, Phosphorylation, etc.)
+    const keyword = ptm.description.split(';')[0].trim();
+    let ptmType: PtmType | undefined = KEYWORD_TO_PTM_TYPE[keyword];
+
+    if (!ptmType) {
+        if (ptm.description.toLowerCase().includes('ubiquitin')) {
+            ptmType = PtmType.Ubiquitination;
+        } else if (ptm.description.toLowerCase().includes('sumo')) {
+            ptmType = PtmType.Sumoylation;
+        }
+    }
+
+    return ptmType || PtmType.Other;
```

Comment:
maybe like this? I feel like `PtmType.Other` probably better to be in `else` logic
```
    if (!ptmType) {
        if (ptm.description.toLowerCase().includes('ubiquitin')) {
            ptmType = PtmType.Ubiquitination;
        } else if (ptm.description.toLowerCase().includes('sumo')) {
            ptmType = PtmType.Sumoylation;
        } else {
            ptmType = PtmType.Other
        }
    }

    return ptmType;
```

Reply:
agree.